src/repo/fs.rs

```diff
@@ -89,3 +90,214 @@ pub fn create_file<P: AsRef<Path>>(path: P) -> io::Result<File> {
         Ok(file) => Ok(file),
     }
 }
+
+#[cfg(unix)]
+const OS_ERROR_DIR_NOT_EMPTY: i32 = 39 /* ENOTEMPTY */;
+#[cfg(windows)]
+const OS_ERROR_DIR_NOT_EMPTY: i32 = 145 /* ERROR_DIR_NOT_EMPTY */;
+
+/// Removes empty directories by starting at [`leaf_dir`] and bubbling up
+/// until `boundary_dir` is reached.
+///
+/// # Arguments
+/// * `leaf_dir` - the leaf directory from which to start the removal
+/// * `boundary_dir` - stops when this directory is reached
+pub fn remove_empty_dirs<P1: AsRef<Path>, P2: AsRef<Path>>(
+    leaf_dir: P1,
+    boundary_dir: P2,
+) -> io::Result<()> {
+    assert!(
+        Iterator::eq(
+            leaf_dir
+                .as_ref()
+                .components()
+                .take(boundary_dir.as_ref().components().count()),
+            boundary_dir.as_ref().components()
+        ),
+        "leaf_dir must be a sub-directory of boundary_dir"
+    );
+    assert!(
+        leaf_dir.as_ref() != boundary_dir.as_ref(),
+        "leaf_dir cannot be the same directory as boundary_dir"
+    );
+
+    let current_dir = leaf_dir.as_ref().to_path_buf();
+    match fs::remove_dir(current_dir) {
+        Err(e) if matches!(e.raw_os_error(), Some(OS_ERROR_DIR_NOT_EMPTY)) => Ok(()),
+        Ok(()) => {
+            let parent_dir = leaf_dir.as_ref().parent().unwrap();
+            if parent_dir != boundary_dir.as_ref() {
```

Comment:
Just checking, what happens if relative paths are involved.

boundary_dir = foo/bar/
parent_dir = foo/bar/baz/../../qux

Is it possible that foo/qux could get deleted even though it exists outside the boundary dir? If so, is there any way to ensure that deletions cannot happen outside of the boundary dir?

The risk seems fairly minimal since this function can only delete empty directories by virtue of fs::remove_dir but thought it worth asking.

Reply:
Could we protect against it with an assertion? I'm thinking, abort if any of the path components are `..`, if we're not expecting there. My thinking here is just to be defensive: I can't immediately see any significant risk here since it is just deletion of empty directories, but it seems a good principle to protect against deletion outside of the boundary directory in the presence of a malformed input.